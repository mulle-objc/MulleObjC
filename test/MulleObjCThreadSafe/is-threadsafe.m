#import <MulleObjC/MulleObjC.h>

@interface Foo : NSObject <MulleObjCThreadUnsafe>
@end


@implementation Foo
@end


@interface Bar : NSObject <MulleObjCThreadSafe>
@end


@implementation Bar
@end


@interface FooBar : Bar <MulleObjCThreadUnsafe>
@end


@implementation FooBar
@end


@interface BarFoo : Foo <MulleObjCThreadSafe>
@end


@implementation BarFoo
@end

// this just inherits Bar and does not modify protocols: threadsafety as Bar
@interface BarKeeper : Bar 
@end


@implementation BarKeeper
@end

static void   test( Class cls)
{
   id  obj;

   mulle_printf( "+[%s mulleIsThreadSafe:] %btd\n",
                  MulleObjCClassGetNameUTF8String( cls),
                  [cls mulleIsThreadSafe]);
   obj = [cls object];
   mulle_printf( "-[%s mulleIsThreadSafe:] %btd\n",
                 MulleObjCClassGetNameUTF8String( cls),
                 [obj mulleIsThreadSafe]);
}


int main( void)
{
   test( [Foo class]);
   test( [Bar class]);
   test( [FooBar class]);
   test( [BarFoo class]);
   test( [BarKeeper class]);

   return( 0);
}