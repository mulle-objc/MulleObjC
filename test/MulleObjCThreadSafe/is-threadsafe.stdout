+[Foo mulleIsThreadSafe:] YES
-[Foo mulleIsThreadSafe:] NO
+[Bar mulleIsThreadSafe:] YES
-[Bar mulleIsThreadSafe:] YES
+[FooBar mulleIsThreadSafe:] YES
-[FooBar mulleIsThreadSafe:] NO
+[BarFoo mulleIsThreadSafe:] YES
-[BarFoo mulleIsThreadSafe:] YES
+[BarKeeper mulleIsThreadSafe:] YES
-[BarKeeper mulleIsThreadSafe:] YES
